src/Disciplina/DB/DSL/Types.hs

```diff
@@ -0,0 +1,100 @@
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE GADTs #-}
+
+module Disciplina.DB.DSL.Types
+       ( Query (..)
+        ,WHERE(..)
+        ,Tx(..)
+        ,Txs(..)
+        ,Obj(..)
+        ,FilterEntity(..)
+        ,FilterExpr(..)
+        ,NodeWithProp(..)
+       ) where
+
+import Universum
+import qualified Disciplina.Core as Core (Grade(..), SubjectId(..))
+
+
+
+class FilterEntity a b where
+  data FilterLHS a b :: * -- LHS of equation
+  type FilterRHS a b :: * -- RHS of equation
+
+
+data Tx = Tx
+data TxHash
+
+instance FilterEntity Tx TxHash where
+  data FilterLHS Tx TxHash = TxHash
+  type FilterRHS Tx TxHash = Text
+
+
+data Txs = Txs
+data TxSubjectId
+data TxGrade
+
+instance FilterEntity Txs TxSubjectId where
+  data FilterLHS Txs TxSubjectId = TxSubjectId
+  type FilterRHS Txs TxSubjectId = Core.SubjectId
+
+instance FilterEntity Txs TxGrade  where
+  data FilterLHS Txs TxGrade = TxGrade
+  type FilterRHS Txs TxGrade = Core.Grade
+
+
+data Obj = Obj
+data ObjHash
+instance FilterEntity Obj ObjHash where
+  data FilterLHS Obj ObjHash = ObjHash
+  type FilterRHS Obj ObjHash = Text
+
+
+
+data NodeWithProp = SubjectNode Core.SubjectId
+                  | TxSubjectNode
+
+data WHERE = WHERE
+
+
+data FilterExpr a b where
+   (:==) :: FilterLHS a b -> FilterRHS a b -> FilterExpr a b
+   (:>=) :: FilterLHS a b -> FilterRHS a b -> FilterExpr a b
+   (:&)  :: FilterExpr a b -> FilterExpr c d -> FilterExpr e f
```

Comment:
If `a` and `c` are different, what's the meaning of the expression? Can you meaningfully combine two expressions, one of which selects `Txs`, and another selects `Obj`? And what meaning we can possibly attribute to `f` then?
With this definition of `FilterExpr` and `Query`, which type `q4 txParams subjId` has? (It has type `Query Txs f`, where `f` is ambiguous and there's no defined type which we can put in place of `f`).

I think that slightly different approach should be introduced here - `FilterExpr` (and, consequently, `Query`) should have only one type parameter, like this:
```
data FilterExpr a where
    (:==) :: forall b. FilterLHS a b -> FilterRHS a b -> FilterExpr a
    (:>=) :: forall b. FilterLHS a b -> FilterRHS a b -> FilterExpr a
    (:&)  :: FilterExpr a -> FilterExpr a -> FilterExpr a
    ...
```

Reply:
Thanks, I will try this approach and see if it compiles. The problem I had when I first tried this (without the `forall b` was that the parameters to `(:&)` both had to be of the same type `FilterExpr a`, but they were not.